openex-api/src/main/java/io/openex/config/AppSecurityConfig.java

```diff
@@ -26,165 +29,398 @@
+        return userService.createUser(createUserInput, 0);
+      } else {
+        // If user exists, update it
+        User currentUser = optionalUser.get();
+        currentUser.setFirstname(firstName);
+        currentUser.setLastname(lastName);
+        if (!rolesAdmin.isEmpty()) {
+          currentUser.setAdmin(isAdmin);
+        }
+        return userService.updateUser(currentUser);
+      }
+    }
+    return null;
+  }
+
+  public User userOauth2Management(OAuth2AccessToken accessToken, ClientRegistration clientRegistration,
+      OAuth2User user) {
+    String emailAttribute = user.getAttribute("email");
+    String registrationId = clientRegistration.getRegistrationId();
+    List<String> rolesFromToken = extractRolesFromToken(accessToken, registrationId);
+    User userLogin = userManagement(emailAttribute, registrationId, rolesFromToken, user.getAttribute("given_name"),
+        user.getAttribute("family_name"));
+
+    if (userLogin != null) {
+      return userLogin;
     }
 
-    @Bean
-    public OAuth2UserService<OidcUserRequest, OidcUser> oidcUserService() {
-        OidcUserService delegate = new OidcUserService();
-        return request -> oauthUserManagement(request.getAccessToken(), request.getClientRegistration(), delegate.loadUser(request));
+    OAuth2Error authError = new OAuth2Error("invalid_token", "User conversion fail", "");
+    throw new OAuth2AuthenticationException(authError);
+  }
+
+  public User userSaml2Management(Saml2AuthenticatedPrincipal user) {
+    String emailAttribute = user.getName();
+    String registrationId = user.getRelyingPartyRegistrationId();
+    List<String> rolesFromUser = extractRolesFromUser(user, registrationId);
+    User userLogin = null;
+    try {
+      userLogin = userManagement(emailAttribute, registrationId, rolesFromUser,
+          user.getFirstAttribute(
+              env.getProperty("openex.provider." + registrationId + ".firstname_attribute_key", String.class, "")),
+          user.getFirstAttribute(
+              env.getProperty("openex.provider." + registrationId + ".lastname_attribute_key", String.class, ""))
+      );
+    } catch (Exception e) {
+      LOGGER.log(Level.SEVERE, e.getMessage(), e);
     }
 
-    @Bean
-    public OAuth2UserService<OAuth2UserRequest, OAuth2User> oauth2UserService() {
-        DefaultOAuth2UserService delegate = new DefaultOAuth2UserService();
-        return request -> oauthUserManagement(request.getAccessToken(), request.getClientRegistration(), delegate.loadUser(request));
+    if (userLogin != null) {
+      return userLogin;
     }
+
+    Saml2Error authError = new Saml2Error("invalid_token", "User conversion fail");
+    throw new Saml2AuthenticationException(authError);
+  }
+
+  public OidcUser oidcUserManagement(OAuth2AccessToken accessToken, ClientRegistration clientRegistration,
+      OAuth2User user) {
+    User loginUser = userOauth2Management(accessToken, clientRegistration, user);
+    return new OpenExOidcUser() {
```

Comment:
OpenExOidcUser can be a class to move all this next line and be more readable
What do you think ?